pkg/controller/direct/apigee/envgroup_controller.go

```diff
@@ -0,0 +1,278 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package apigee
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+	"time"
+
+	krm "github.com/GoogleCloudPlatform/k8s-config-connector/apis/apigee/v1alpha1"
+	refs "github.com/GoogleCloudPlatform/k8s-config-connector/apis/refs/v1alpha1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/config"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/directbase"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/registry"
+
+	api "google.golang.org/api/apigee/v1"
+	"google.golang.org/protobuf/types/known/fieldmaskpb"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	serviceDomain = "//apigee.googleapis.com"
+	ctrlName      = "apigee-envgroup-controller"
+)
+
+func init() {
+	registry.RegisterModel(krm.ApigeeEnvgroupGVK, NewApigeeEnvgroupModel)
+}
+
+func NewApigeeEnvgroupModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+	return &modelApigeeEnvgroup{config: config}, nil
+}
+
+var _ directbase.Model = &modelApigeeEnvgroup{}
+
+type modelApigeeEnvgroup struct {
+	config *config.ControllerConfig
+}
+
+func (m *modelApigeeEnvgroup) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	gcpClient, err := newGCPClient(ctx, m.config)
+	if err != nil {
+		return nil, err
+	}
+
+	apigeeEnvgroupClient, err := gcpClient.newApigeeEnvgroupClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	obj := &krm.ApigeeEnvgroup{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	id, err := krm.NewGoogleCloudApigeeV1EnvironmentGroupIdentity(ctx, reader, obj)
+	if err != nil {
+		return nil, err
+	}
+
+	mapCtx := &direct.MapContext{}
+	desired := ApigeeEnvgroup_ToApi(mapCtx, obj)
+	if mapCtx.Err() != nil {
+		return nil, mapCtx.Err()
+	}
+
+	return &Adapter{
+		id:                   id,
+		desired:              desired,
+		apigeeEnvgroupClient: apigeeEnvgroupClient,
+	}, nil
+}
+
+func (m *modelApigeeEnvgroup) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type Adapter struct {
+	id                   *krm.GoogleCloudApigeeV1EnvironmentGroupIdentity
+	desired              *api.GoogleCloudApigeeV1EnvironmentGroup
+	actual               *api.GoogleCloudApigeeV1EnvironmentGroup
+	apigeeEnvgroupClient *apigeeEnvgroupClient
```

Comment:
Suggest splitting apigeeEnvgroupClient to gcp client and operator client.  It is more readable in the following code (create, waitforOpt) because for KCC controller they represents two different kinds of services and HTTP calls.